cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppVersionUpgradeService.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.services;
+
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.internal.app.runtime.schedule.store.DatasetBasedStreamSizeScheduleStore;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import com.google.protobuf.Service;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * {@link Service} that will start the upgrade threads of Datasets and also provides the status of upgrade.
+ */
+public class AppVersionUpgradeService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(AppVersionUpgradeService.class);
+
+  private final DatasetBasedStreamSizeScheduleStore streamSizeScheduleStore;
+  private final ExecutorService executorService;
+
+  @Inject
+  AppVersionUpgradeService(DatasetBasedStreamSizeScheduleStore streamSizeScheduleStore) {
+    this.streamSizeScheduleStore = streamSizeScheduleStore;
+    this.executorService = Executors.newFixedThreadPool(
+      3, Threads.createDaemonThreadFactory("app-version-upgrade-thread-%d"));
```

Comment:
We will be adding two more threads for DatasetBasedTimeScheduleStore upgrade and AppMetadataStore upgrade. I have reduced it to 1 for now, will bump it up in subsequent commits.